### DISCUSSION TITLE
Tycho failed to resolve target definition file for zipped P2 repository if authentication is required

### DISCUSSION THREAD

**Discussion by mfsoliveira** [eclipse-tycho/tycho#1191] [Q&A] [answered]
I am trying to use a zipped P2 repository, which was uploaded into a private NEXUS server inside a RAW repository.
I am getting the following error: 

` Failed to resolve target definition file:/D:/path/to/target/file.target: Failed to load p2 metadata repository from location jar:https://server.url/repository/raw-p2-private-repository/plugin_dependency/latest/plugin_dependency.zip!/: Unable to read repository at: jar:https://server.url/repository/raw-p2-private-repository/plugin_dependency/latest/plugin_dependency.zip!/site.xml. Too many failed login attempts.`

What we already tested (Thanks @cdietrich for additional tests) :
- Referring to  a local path like  <repository  location="jar:file//d:/path/to/plugin_dependency.zip!/"/> is working fine
- Referring to the same P2 repository in the same server but unzip is working fine
- Referring to a zipped P2 location without authentication is working fine.
- Referring to a zipped P2 location using a local server is failing with the same error message

I am using Java 11 and tycho version 2.7.4

The server is configured in settings.xml like that:

```
<servers>
  <server>
    <id>artifacts.server</id>
    <username>${env.NEXUS_USER}</username>
    <password>${env.NEXUS_PASS}</password>
  </server>
</servers>
```

I have the repository defined in my configuration POM and it uses the same id used in settings.xml:

```
<repositories>
  <repository>
    <id>artifacts.server</id>
    <url>https://server.url/repository/raw-p2-private-repository/plugin_dependency/latest</url>
    <layout>p2</layout>
  </repository>
</repositories>
```
Finally, the repository is defined in the target platform file:
```
<location includeAllPlatforms="false" includeConfigurePhase="false" includeMode="planner" includeSource="true" type="InstallableUnit">
  <unit id="com.required.unit.feature.group" version="1.2.3"/>	
  <repository id="artifacts.server" location="jar:https://server.url/repository/raw-p2-private-repository/plugin_dependency/latest/plugin_dependency.zip!/"/>
  </location>
</locations>
```

The above setup is based on [Tycho - Target_Platform](https://wiki.eclipse.org/Tycho/Target_Platform#Layout_p2) and [Tycho - Authentication and Mirrors](https://wiki.eclipse.org/Tycho/Target_Platform/Authentication_and_Mirrors) documentation. In special the following paragraph:

> When accessing a p2 repository, Tycho will check if there is a  <server> entry with matching ID and use these credentials. The ID of a p2 repository can be specified explicitly, e.g. when the repository is [added in the POM](https://wiki.eclipse.org/Tycho/Target_Platform#Layout_p2). If no ID is specified, e.g. in a target file, the repository URL (without trailing slash) is used as ID (since Tycho 0.16.0).

**Comment by cdietrich** [comment]
i did some remote debugging and saw the the `RemoteRepositoryLoadingHelper.certainlyNoRemoteURL` may return true as the uri will be opaque

```
public class Main2 {

	public static void main(String[] args) throws URISyntaxException {
		URI uri = new URI("jar:http://localhost:8008/org.eclipse.xtext.sdk.p2-repository-2.28.0-SNAPSHOT.zip!/");
		System.out.println(uri.isOpaque());

	}

}
```
so the code that sets up auth may not be called at all

**Comment by laeubi** [comment, marked as answer]
1. What version of Tycho are you using?
2. Please note that your current approach will download the file **multiple times**, P2 do not support zipped P2 sites nor does PDE
3. You may better consume it as a regular maven dependecy with `mvn:<groupid>:<artifactid>:<version>:zip` instead

**Reply by cdietrich** [reply]
problem can be seen in tycho 2.7.4 and 3.0.0-SNAPSHOT
3 wont work in eclipse, pde supports this uri scheme with jar:http ....

**Reply by laeubi** [reply]
> 3 wont work in eclipse, pde supports this uri scheme with jar:http ....

PDE support URIs, but it will download the whole zip for every single file inside the ZIP ... if you use jar:http.

With m2e, PDE support mvn protocol in the IDE as well.

**Reply by cdietrich** [reply]
so it is still better to unpack the update site on the server

**Reply by laeubi** [reply]
> so it is still better to unpack the update site on the server

No it is better to simply use the `mvn:` protocol ... ;-)

**Reply by cdietrich** [reply]
but then you cannot select what to take from there, correct, as you can do with .target

**Reply by laeubi** [reply]
> but then you cannot select what to take from there, correct, as you can do with .target

You cannot do what? As described you can simply use this as an URL like you can use for http ... and then select the desired units.

**Reply by cdietrich** [reply]
so i put this url to the target file and not the pom?

**Reply by mfsoliveira** [reply]
> 

So it should look like the location bellow, according to [Tycho p2 Repository Plugin](https://www.eclipse.org/tycho/sitedocs/tycho-p2/tycho-p2-repository-plugin/plugin-info.html)
```
<location includeAllPlatforms="false" includeConfigurePhase="true" includeMode="planner" includeSource="true" type="InstallableUnit">
<repository location="mvn:[grouId]:[artifactId]:[version]:zip:p2site"/>
-- list desired units here --
</location>
```

but should the p2 repository also be created using [tycho-p2-repository:assemble-maven-repository] (https://www.eclipse.org/tycho/sitedocs/tycho-p2/tycho-p2-repository-plugin/assemble-maven-repository-mojo.html)?

I have the unzipped P2 repository. We reached this problem, just because uploading the zipped file was easier for me than the complete p2, for which I needed to create a POM instead of a simple curl command. After the problem, I pushed forward mixing curiosity and a possible bug hunting as initially suggested by Christian.

**Reply by laeubi** [reply]
> so i put this url to the target file and not the pom?

Yes

> So it should look like the location bellow,

No it depends how you deploy it, a zip file can be deployed to nexus under whatever GAV(C) you want

> but should the p2 repository also be created using

This is a special use-case where you do *not* use Tycho to produce the update-site.

> was easier for me than the complete p2, for which I needed to create a POM instead of a simple curl command

Deploy a file does not require a pom, you can use a arbitrary file and specify the coordinates on the command line see https://maven.apache.org/plugins/maven-deploy-plugin/usage.html#the-deploy-deploy-file-mojo

This has the advantage that you fully can reuse maven workflows, settings and so on ...
